### PR TITLE
Deps/opmsg

### PR DESCRIPTION
**Pull request by fewf**
Hi @ry -- this is following up on https://github.com/denoland/deno/pull/2223

Definitely not finished, but you can let me know if it's going in the right direction. I hacked `modules` into the `ThreadSafeState` object, so it can compile, but it's not behaving--I'm sure for obvious reasons. I'll leave some more comments around the code. Also thinking we'll need to add in some tests somewhere/

### CODE REVIEW

File: cli/msg.fbs [hunk 1]
```diff
@@ -279,6 +281,12 @@ table EnvironRes {
   map: [KeyValue];
 }
 
+table Deps {}
+
+table DepsRes {
+  data: [ubyte];
```
Comment (fewf) [inline cli/msg.fbs]: wasn't sure if this should be type `string` or not, since it could be arbitrarily large

File: cli/ops.rs [hunk 2]
```diff
@@ -620,6 +622,46 @@ fn op_env(
...
+    &state,
+    &state.main_module().unwrap(),
+    ".",
+  );
+  if let Err(e) = maybe_out {
+    return odd_future(e);
+  }
+  let out = maybe_out.unwrap();
+
+  if let Some(deps) = state.modules.deps(&out.module_name) {
```
Comment (fewf) [inline cli/ops.rs]: this does not work right now

File: js/deps.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import * as msg from "gen/cli/msg_generated";
+import { sendSync } from "./dispatch";
+import { decodeMessage } from "./workers";
+import { assert } from "./util";
+import * as flatbuffers from "./flatbuffers";
+
+export function deps(): any {
```
Comment (fewf) [inline js/deps.ts]: def want to come back and drop this `any` return type/create Interface for deps, etc

File: cli/ops.rs [hunk 4]
```diff
@@ -620,6 +622,46 @@ fn op_env(
...
+  state: &ThreadSafeState,
+  base: &msg::Base<'_>,
+  data: Option<PinnedBuf>,
+) -> Box<OpWithError> {
+  assert!(data.is_none());
+  let cmd_id = base.cmd_id();
+
+  let maybe_out = worker::fetch_module_meta_data_and_maybe_compile(
+    &state,
+    &state.main_module().unwrap(),
```
Comment (bartlomieju) [inline cli/ops.rs]: I think you can easily take it from message to allow to inspect any file:
```
Deno.deps("https://deno.land/log/mod.ds");
```

Comment (ry) [reply cli/ops.rs]: > I think you can easily take module name from message to allow to inspect any file:

I'm thinking about it, but it's not obvious to me how to find out which module as called an op... It seems this is some functionality we need in core.

File: cli/state.rs [hunk 5]
```diff
@@ -65,6 +66,7 @@ pub struct State {
   pub start_time: Instant,
   pub resource: resources::Resource,
   pub dispatch_selector: ops::OpSelector,
+  pub modules: deno::Modules,
```
Comment (ry) [inline cli/state.rs]: This is something we need to do. However it needs to also be removed from the Worker struct. I will prepare a PR which does just this.

**Comment by ry** [comment]
After some research, it doesn't appear V8 gives us much in the way of determining which module called a function. 

One way this might be accomplished is by manually traversing the current stack to figure out 
It seems this *could* be determined is by calling [v8::StackTrace::CurrentStackTrace()](https://github.com/denoland/deno_third_party/blob/b9286ced4a1ee9d8c9e58088803aa3c0d60011c7/v8/include/v8.h#L1827-L1835). It's not something we could do for every op, for performance reasons, but maybe it could be called during `deps()` or other ops that need it. Maybe something like `const char* deno_calling_module(Deno* d)` ...

**Comment by fewf** [comment]
Thanks for the info @ry. I think I understand the issue--it's that if A has B for a dependency and B calls `deps`, it should return only B's dependencies, right? That does seem tricky. I notice `console.trace` is not implemented in Deno--is that a todo? There's a way to get a stack trace like so (cribbed from https://stackoverflow.com/a/28118170 ):
```
  let stack;

  try {
    throw new Error('');
  }
  catch (error) {
    stack = error.stack || '';
  }
```

And that seems to work in Deno.

For the moment, I've pushed up some more proposed changes to better Typescript-ify the `deps` function and add the option to return an object instead of an array mentioned here https://github.com/denoland/deno/issues/2096 (altho maybe that functionality should be elsewhere). I tried to rely on conventions I could find elsewhere.

File: cli/state.rs (hunk 5 above)
Comment (bartlomieju) [reply cli/state.rs]: `Modules` had been moved to `State` - @fewf you can merge latest changes from `master`

**Comment by ry** [comment]
@fewf I'm closing stale PRs and I'm afraid this one qualifies. We've done a lot of refactoring in modules recently and I imagine this would need some strenuous rebasing. I'm going to close this - feel free to hit me up on gitter and discuss if you want to pursue. Sorry for not getting it in sooner - but there were some open questions.

**Pull request closed without merging**